Write a program to find the average of 1000 random digits 0, 1, 2, 3, 4, 5, 6, 7, 8, or 9. Have the program test to see if the average lies within three standard deviations of the expected value of 4.5. Modify the program so that it repeats this simulation 1000 times and keeps track of the number of times the test is passed. Does your outcome agree with the Central Limit Theorem? Use Python.

We'll start with some setup, establishing key parameters. Most are directly given (the number of simulations, the size of each sample, the expected value), but we must calculate a few.

Per the CLT, the standard deviation of our sample distribution of the mean (aka the standard error) is defined as $$ \sigma / \sqrt{n} $$. In this case, our sample standard deviation ($$\sigma$$) is defined by the standard deviation of a uniform distribution, i.e. $$(b - a)^2/12$$, where $$a$$ and $$b$$ are the lower and upper bounds of the distribution, respectively. We can use this standard error to developer the +/- 3 standard deviation range, within which we expect our sample mean to fall ~99.7% of the time (per the CLT).

In [1]:
import numpy as np
np.random.seed(412)

n_sims = 1000
sample_size = 1000

exp_value = 4.5
std_dev = np.sqrt( (9 - 0)**2 / 12 )
range_3sd = [
    exp_value - 3 * std_dev / np.sqrt(sample_size), 
    exp_value + 3 * std_dev / np.sqrt(sample_size)
]

print(f'Range: {range_3sd[0]:.4f} to {range_3sd[1]:.4f}')

Range: 4.2535 to 4.7465


With these parameters established, we can run our simulation. In each iteration, we generate 1000 random integers from 1 to 9 (note that np.randint() has an exclusive upper bound). With those 1000 integers, we calculate the mean, then check whether it falls within our +/- 3 standard deviation range. If so, we return a 1 for pass, and if not, we return 0. We can then calculate the proportion of passes (i.e. 1s) to total simulations (1000).

In [2]:
results = []

for sim in range(n_sims):
    sample = np.random.randint(0, 10, sample_size)
    avg = sample.mean()
    range_check = 1 if range_3sd[0] <= avg <= range_3sd[1] else 0
    results.append(range_check)
    
print(
    f'# within +/- 3SD: {sum(results)}\n'
    f'% within +/- 3SD: {sum(results)/n_sims*100:.2f}%'
)

# within +/- 3SD: 996
% within +/- 3SD: 99.60%


Our results support the CLT! We see that, in 99.6% of the simulations, the sample mean falls within the +/- 3SD range, which is very close to the 99.7% expectation. 